### PR TITLE
implemented multi user sign in

### LINKED ISSUE #33: User sign in / sign out / sign up
@sujaykumarreddy 

- User privileges assigned as per the actions
- Multi user
  - Organiser
  - Speaker
  - Admin
  - Stakeholder
- As less as urls as can be
- model design - @tapaswenipathak posts

### PR DESCRIPTION
**Pull request by sujay0399**
fix  #33

### CODE REVIEW

**Review by tapaswenipathak — COMMENTED** [review]
As discussing, instead of https://github.com/tapaswenipathak/CFP/issues/33#issuecomment-495069252, let's implement https://github.com/tapaswenipathak/CFP/issues/33#issuecomment-495071043. I am adding inputs.

File: yourcfp/users/models.py [hunk 1]
```diff
@@ -1,3 +1,13 @@
+
+from django.contrib.auth.models import AbstractUser
 from django.db import models
 
-# Create your models here.
+class User(AbstractUser):
```
Comment (tapaswenipathak) [inline yourcfp/users/models.py:6]: Let's do  `class User(AbstractUser):`
   `User_Type_Choices` - list choices

- `UserType`

- Create a class Role(models.model)

Comment (tapaswenipathak) [reply yourcfp/users/models.py]: Roles

- Admin
- Organisers
- Reviewers
- Speakers
- Sponsors
- Volunteers

File: yourcfp/users/models.py [hunk 2]
```diff
@@ -1,3 +1,13 @@
...
+    is_speaker = models.BooleanField(default=False)
+    is_organizer = models.BooleanField(default=False)
+
+class Speaker(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+
+class Organizer(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
```
Comment (tapaswenipathak) [inline yourcfp/users/models.py:22]: Roles

- Admin
- Organisers
- Reviewers
- Speakers
- Sponsors
- Volunteers

File: yourcfp/yourcfp/urls.py [hunk 3]
```diff
@@ -20,5 +20,8 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('', include('home.urls')),
-    path('register/', user_views.RegisterView.as_view(), name='register')
+    path('register-as-organizer/', user_views.OrganizerRegisterView.as_view(), name='register-as-organizer'),
+    path('register-as-speaker/', user_views.SpeakerRegisterView.as_view(), name='register-as-speaker'),
+    path('speaker_success/', user_views.sp_success, name='speaker-success'),
+    path('organizer_success/', user_views.or_success, name='organizer-success')
```
Comment (tapaswenipathak) [inline yourcfp/yourcfp/urls.py]: Let's write url per https://github.com/tapaswenipathak/CFP/issues/33#issuecomment-495071043

File: yourcfp/users/decorators.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    actual_decorator = user_passes_test(
+        lambda u: u.is_active and u.is_organizer,
+        login_url=login_url,
+        redirect_field_name=redirect_field_name
+    )
+    if function:
+        return actual_decorator(function)
+    return actual_decorator
+
+
```
Comment (tapaswenipathak) [inline yourcfp/users/decorators.py:19]: Login decorator would be per

- admin
- organiser (reviewers, sponsors, volunteers)
- speaker

**Review by tapaswenipathak — COMMENTED** [review]
Should we do a POC having

- https://github.com/Nitron/django-cas-provider + https://github.com/Nitron/django-cas-consumer

?

Django single sign on would be cool if we have in the app. I will write diagram or will make a poc.

**Review by tapaswenipathak — COMMENTED** [review]
Table would be `User` table. Everything else would be a foreign key pointing `User` table.

# User Table

- id (primary key)
- role (any of the roles I wrote)
- roles_id(foreign key)

# Per Role table
- id (primary key + foreign key)
- role information

# Proposal table

**Review by tapaswenipathak — COMMENTED** [review]
Organiser + Proposal

Organiser Table

- User table foreign key
- User table role
- Proposal_id (primary key, foreign key)

**Review by tapaswenipathak — COMMENTED** [review]
CRUD operations in urls would be ok

https://github.com/tapaswenipathak/CFP/issues/33#issuecomment-495071043

**Review by tapaswenipathak — COMMENTED** [review]
Let's designate user table as master table.

File: yourcfp/users/forms.py [hunk 5]
```diff
@@ -1,10 +1,32 @@
...
-        fields = ['username', 'email', 'first_name', 'last_name', 'password1', 'password2']
+        fields = ['username', 'email', 'password1', 'password2']
+
+    @transaction.atomic
+    def save(self):
+        user = super().save(commit=False)
+        user.is_speaker = True
+        user.save()
+        student = Speaker.objects.create(user=user)
+        return user
```
Comment (tapaswenipathak) [inline yourcfp/users/forms.py:32]: Field defining - @tapaswenipathak

File: yourcfp/users/views.py [hunk 6]
```diff
@@ -1,12 +1,34 @@
...
+    success_url = reverse_lazy('organizer-success')
+
+    def get_context_data(self, **kwargs):
+        kwargs['user_type'] = 'organizer'
+        return super().get_context_data(**kwargs)
+
+class SpeakerRegisterView(CreateView):
+    model = User
+    form_class = SpeakerSignUpForm
+    template_name = 'users/speaker_register_form.html'
```
Comment (tapaswenipathak) [inline yourcfp/users/views.py:25]: would be scraped.

File: yourcfp/users/views.py [hunk 7]
```diff
@@ -1,12 +1,34 @@
...
 # Create your views here
 
-class RegisterView(CreateView):
-    form_class = UserRegistrationForm
-    template_name = 'users/user_register_form.html'
-    success_url = reverse_lazy('home:index')
+class OrganizerRegisterView(CreateView):
+    model = User
+    form_class = OrganizerSignUpForm
+    template_name = 'users/organizer_register_form.html'
```
Comment (tapaswenipathak) [inline yourcfp/users/views.py:15]: scraped as well.

**Review by tapaswenipathak — COMMENTED** [review]
We will have `users/users_signup.html` in views.

**Review by tapaswenipathak — COMMENTED** [review]
Apps

- Organiser
- Events
- Proposals
- Authentication
- Home
- Administrator
- Volunteers
- Shifts

File: yourcfp/users/models.py (hunk 2 above)
Comment (sujay0399) [reply yourcfp/users/models.py:22]: @tapaswenipathak 
Do we really need to create separate roles, we can create tables and assign permissions.
Reviewers, Volunteers can be relates to organizer table.
Sponsors can be related to event model.

**Comment by sujay0399** [comment]
@tapaswenipathak 
Just implemented profiles model.
Will start working on events and proposals.

**Review by tapaswenipathak — COMMENTED** [review]
@sujaykumarreddy assign me this pr when you are done.

**Comment by sujay0399** [comment]
@tapaswenipathak Implemented forms for submitting proposals.
Will start working on saving them as draft and work on permissions.

**Review by tapaswenipathak — COMMENTED** [review]
Create separate PR for more tickets. I am writing few review comments, review comments and merge.

File: yourcfp/events/forms.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from django import forms
+from django.template.defaultfilters import slugify
+
+from .models import Conference
+
+class ConferenceForm(forms.ModelForm):
+    class Meta:
+        model = Conference
+        fields = ['name', 'description', 'start_date', 'end_date', 'twitter_id', 'venue']
```
Comment (tapaswenipathak) [inline yourcfp/events/forms.py:9]: https://github.com/tapaswenipathak/CFP/issues/42

File: yourcfp/events/models.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    organizer = models.ForeignKey(User, on_delete=models.CASCADE)
+    name = models.CharField(max_length=255)
+    description = models.TextField(default="")
+    start_date = models.DateField()
+    end_date = models.DateField()
+    venue = models.TextField(default="")
+    twitter_id = models.CharField(max_length=100, blank=True, null=True, default='')
+
+    def __str__(self):
+        return self.name
```
Comment (tapaswenipathak) [inline yourcfp/events/models.py:23]: https://github.com/tapaswenipathak/CFP/issues/42

File: yourcfp/events/templates/events/conference_detail.html [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<h1>organizer name : {{ object.organizer }}</h1>
+<p>conference name : {{ object.name }}</p>
+<p>description : {{ object.description }}</p>
+<p>start date : {{ object.start_date|date }}</p>
+<p>end date : {{ object.end_date|date }}</p>
+<p>venue : {{ object.venue }}</p>
+<p>twitter_id : {{ object.twitter_id }}</p>
+<br>
+<a href="{% url 'events:conference-update' object.id %}">update view</a>
```
Comment (tapaswenipathak) [inline yourcfp/events/templates/events/conference_detail.html]: We need a designer for this or I only will design UI as well.

File: yourcfp/events/templates/events/conference_list.html [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<h2>Conferences</h2>
+<ul>
+    {% for conference in object_list %}
+        <li><a href="{% url 'events:conference-detail' conference.id %}">{{ conference.name }}</a></li>
+    {% endfor %}
+</ul>
```
Comment (tapaswenipathak) [inline yourcfp/events/templates/events/conference_list.html:6]: UI design.

File: yourcfp/home/templates/home/home.html [hunk 12]
```diff
@@ -1,29 +1,12 @@
...
-        <h1 class="index">Your Conference Proposal</h1>
-        <p>{% trans "" %}</p>
-        {% if error %}
-            <div class="alert alert-danger" role="alert">
-              You will need to have atleast one organisation to Sign up
-            </div>
-        {% endif %}
-    </div>
-{% endblock %}
+<a href={% url 'users:register-as-speaker' %}>register as speaker</a>
```
Comment (tapaswenipathak) [inline yourcfp/home/templates/home/home.html:1]: UI design.

File: yourcfp/proposals/templates/proposals/proposal.html [hunk 13]
```diff
@@ -0,0 +1,4 @@
+<form method="post">{% csrf_token %}
+    {{ form.as_p }}
+    <input type="submit" value="Save">
+</form>
```
Comment (tapaswenipathak) [inline yourcfp/proposals/templates/proposals/proposal.html]: `csrf_token` check should be present for every implementation.

File: yourcfp/proposals/templates/proposals/proposal_list.html [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<h2>Proposals</h2>
+<ul>
+    {% for proposal in object_list %}
+        <li><a href="#">{{ proposal.author }}'s {{ proposal.name }} proposal </a></li>
+    {% endfor %}
+</ul>
```
Comment (tapaswenipathak) [inline yourcfp/proposals/templates/proposals/proposal_list.html:6]: UI design.

File: yourcfp/users/forms.py (hunk 5 above)
Comment (tapaswenipathak) [reply yourcfp/users/forms.py:32]: https://github.com/tapaswenipathak/CFP/issues/42

**Review by tapaswenipathak — CHANGES_REQUESTED** [review]
Left few minor review comments, let's merge after permissions handling.

File: yourcfp/proposals/templates/proposals/proposal.html (hunk 13 above)
Comment (tapaswenipathak) [reply yourcfp/proposals/templates/proposals/proposal.html]: UI design.

**Pull request closed without merging**